resources/type-system/inference.md

```diff
@@ -1324,51 +1373,106 @@ determined as follows:
-    `v`. _Since `o_2` is an instance of `Future<T>`, by future soundness, `v`
-    must be an instance of `T`. This satisfies expression soundness, since `v`
-    is an instance of `T` , and `T` is the static type._
+- If `T_1` is incompatible with await (as defined in the _extension types_
+  specification), then there is a compile-time error.
+
+- Let `T_2` be `flatten(T_1)`.
+
+- Let `m_2` be `@AWAIT_WITH_TYPE_CHECK<T_2>(m_1)`, with static type
+  `Future<T_2>`.
+
+  - _Note that in many circumstances, it will be trivial for the compiler to
+    establish that `m_1` always evaluates to an instance `v` that satisfies type
+    `Future<T_2>`, in which case `m_2` can be optimized to
+    `@PROMOTED_TYPE<Future<T_2>>(m_1)`._
+
+  - _For soundness, we must prove that whenever
+    `@AWAIT_WITH_TYPE_CHECK<T_2>(m_1)` evaluates `m_1` to a value `v` that is
+    __not__ an instance satisfying type `Future<T_2>`, that `v` __is__ an
+    instance satisfying type `T_2`. This is necessary to ensure that the future
+    created by `@AWAIT_WITH_TYPE_CHECK` will complete to a value satisfying its
+    type signature. Note that `v` is guaranteed to be an instance satisfying
+    type `T_1` (because `T_1` is the static type of `m_1`). So we can establish
+    soundness by assuming that `v` is an instance satisfying type `T_1` and not
+    an instance satisfying type `Future<T_2>`, and then considering two cases:_
+
+    - _If the runtime value of `v` is `null`, then by soundness, `T_1` must be
+      of the form `Null`, `dynamic`, `S*`, or `S?`. Considering each of these:_
+
+      - _If `T_1` is of the form `Null` or `dynamic`, then by the definition of
+        `flatten`, `T_2` must be the same as `T_1`. Therefore, `v` is an
+        instance satisfying type `T_2`, so soundness is satisfied._
+
+      - _If `T_1` is of the form `S*` or `S?`, then by the definition of
+        `flatten`, `T_2` must be of the form `flatten(S)*` or `flatten(S)?`,
+        respectively. `null` is an instance satisfying all types ending in `*`
+        and `?`, so soudness is satisfied._
+
+    - _Otherwise, we need to show that if `v` is a non-null instance satisfying
+      type `T_1`, but not an instance satisfying type `Future<T_2>`, then `v` is
+      an instance satisfying type `T_2`._
+
+    - _Substituting in the definition of `T_2`, we need to show that if `v` is a
+      non-null instance satisfying type `T_1`, but not an instance satisfying
+      type `Future<flatten(T_1)>`, then `v` is an instance satisfying type
+      `flatten(T_1)`. We can prove this by induction on `T_1`:_
+
+      - _If `T_1` is `S?`, then `flatten(T_1)` is `flatten(S)?`. We need to show
+        that if `v` is a non-null instance satisfying type `S?`, but not an
+        instance satisfying type `Future<flatten(S)?>`, then `v` is an instance
+        satisfying type `flatten(S)?`. Assuming `v` is a non-null instance
```

Comment:
@fishythefish similarly here:  if `v` has runtime type `T` and `T <: S?` then either `T` is `Null` or `T <: S`.

Reply:
The union type rules are that `T <: S?` implies `T <: S` and/or `T <: Null`. 
It happens that today `T <: Null` implies `T = Null` or `T <: Never`, and `T <: Never` is subsumed by `T <: S`. 
Or actuall, that `T = Null` up to **Norm**. `T` *can* also be `Never?` or `Null?`. The *flatten* function is defined on *non-normalized* type terms. (We haven't said otherwise, although I'm not *sure* it behaves differently for any different norm-equivalent types. I'm also not completely sure it doesn't.)

It's better to just not make the assumption unless it's needed. If `T <: Null` is enough, then just use that.